src/tools/tidy/src/x_version.rs

```diff
@@ -0,0 +1,80 @@
+use semver::Version;
+use serde_json::Value;
+use std::io::ErrorKind;
+use std::process::{Command, Stdio};
+
+pub fn check(bad: &mut bool) {
+    let result = Command::new("x").arg("--wrapper-version").stdout(Stdio::piped()).spawn();
+    // This runs the command inside a temporary directory.
+    // This allows us to compare output of result to see if `--wrapper-version` is not a recognized argument to x.
+    let temp_result = Command::new("x")
+        .arg("--wrapper-version")
+        .current_dir(std::env::temp_dir())
+        .stdout(Stdio::piped())
+        .spawn();
+
+    let (child, temp_child) = match (result, temp_result) {
+        (Ok(child), Ok(temp_child)) => (child, temp_child),
+        (Err(e), _) | (_, Err(e)) => match e.kind() {
+            ErrorKind::NotFound => return,
+            _ => return tidy_error!(bad, "failed to run `x`: {}", e),
+        },
+    };
+
+    let output = child.wait_with_output().unwrap();
+    let temp_output = temp_child.wait_with_output().unwrap();
+
+    if output != temp_output {
+        return tidy_error!(
+            bad,
+            "Current version of x does not support the `--wrapper-version` argument\nConsider updating to the newer version of x by running `cargo install --path src/tools/x`"
+        );
+    }
+
+    if output.status.success() {
+        let version = String::from_utf8_lossy(&output.stdout);
+        let version = Version::parse(version.trim_end()).unwrap();
+
+        if let Some(expected) = get_x_wrapper_version() {
+            if version < expected {
+                return tidy_error!(
+                    bad,
+                    "Current version of x is {version}, but the latest version is {expected}\nConsider updating to the newer version of x by running `cargo install --path src/tools/x`"
+                );
+            }
+        } else {
+            return tidy_error!(
+                bad,
+                "Unable to parse the latest version of `x` at `src/tools/x/Cargo.toml`"
+            );
+        }
+    } else {
+        return tidy_error!(bad, "failed to check version of `x`: {}", output.status);
+    }
+}
+
+// Parse latest version out of `x` Cargo.toml
+fn get_x_wrapper_version() -> Option<Version> {
+    let cmd = Command::new("cargo")
```

Comment:
Do we know this will be run from the root of the rust-lang/rust checkout? 
I tested this by being in some random directory (inside the project) & it did use the root, but can we count on that or should the directory be set?

Reply:
You can test by running `../../../x.py test tidy` from `src/tools/x`, which isn't in the workspace. If that works, and `rust/x.py test tidy` works (from outside the source directory), this is fine :)

Reply:
err sorry - actually can you use `cargo_metadata`, like in https://github.com/rust-lang/rust/blob/e9ca6636e108cb4380d38a45c5ebb1d485079210/src/tools/tidy/src/deps.rs#L347-L354? that takes `root` as an argument which avoids this question altogether.

Reply:
Running both of the above commands didn't end up working for me, even with the new change to use `cargo_metadata`.

I end up getting tidy errors for autogenerated files in the build directory...
```
Building rustbuild
    Finished dev [unoptimized] target(s) in 0.08s
Building stage0 tool tidy (aarch64-apple-darwin)
   Compiling tidy v0.1.0 (/Users/jhaigh/developer/Rust/rust/src/tools/tidy)
    Finished release [optimized] target(s) in 2.62s
tidy check
* 633 error codes
* highest error code: E0791
tidy error: /Users/jhaigh/developer/Rust/rust/src/tools/x/build/bootstrap/debug/build/typenum-
949154f977098d36/out/consts.rs: leading newline
tidy error: /Users/jhaigh/developer/Rust/rust/src/tools/x/build/bootstrap/debug/build/typenum-
949154f977098d36/out/consts.rs: missing trailing newline
tidy error: /Users/jhaigh/developer/Rust/rust/src/tools/x/build/bootstrap/debug/build/typenum-
949154f977098d36/out/tests.rs: leading newline
tidy error: /Users/jhaigh/developer/Rust/rust/src/tools/x/build/bootstrap/debug/build/typenum-
949154f977098d36/out/tests.rs: missing trailing newline
tidy error: /Users/jhaigh/developer/Rust/rust/src/tools/x/build/bootstrap/debug/build/typenum-
949154f977098d36/out/tests.rs: too many lines (17255) (add `// ignore-tidy-filelength` to the 
file to suppress this error)
tidy error: /Users/jhaigh/developer/Rust/rust/src/tools/x/build/bootstrap/debug/build/typenum-
949154f977098d36/out/op.rs: leading newline
tidy error: /Users/jhaigh/developer/Rust/rust/src/tools/x/build/bootstrap/debug/build/typenum-
949154f977098d36/out/op.rs: missing trailing newline
tidy error: /Users/jhaigh/developer/Rust/rust/src/tools/x/build/aarch64-apple-darwin/stage0/sh
are/doc/cargo/README.md: too many trailing newlines (2)
tidy error: /Users/jhaigh/developer/Rust/rust/src/tools/x/build/aarch64-apple-darwin/stage0/sh
are/doc/rustfmt/README.md:108: trailing whitespace
* 392 features
some tidy checks failed
Build completed unsuccessfully in 0:00:05
```

Reply:
ah yeah that's from when I asked you to test `../../../x test tidy`, sorry about that 😅 `rm -rf src/tools/x/build` will make the errors go away (although I've been wanting to get tidy to ignore untracked files for a while; but that's a separate PR.)